File: Pendulum.m
function state = Pendulum(x_pos0,y_pos0,z_pos0,vx0,vy0,vz0)
global L  g  b  m dt
% Initial conditions [x, x_dot, y, y_dot, z, z_dot]
% x_pos0 = L/sqrt(2);  % Initial x-position
% vx0 = 0;  % Initial x-velocity
% y_pos0 = L/sqrt(2);  % Initial y-position
% vy0 = 0;  % Initial y-velocity
% z_pos0 = 0; % Initial z-position (length of the pendulum)
% vz0 = 0;  % Initial z-velocity

% Time parameters
     
t_end = 3;    % End time for simulation
t = 0:dt:t_end; % Time vector

% Preallocate arrays for storing results

v_dot = zeros(length(t),3);
p_dot = zeros(length(t),3);
p  = zeros(length(t),3);

% Set initial conditions
p(1,1) = x_pos0;
p(1,2) = y_pos0;
p(1,3) = z_pos0;
p_dot(1,1) = vx0;
p_dot(1,2) = vy0;
p_dot(1,3) = vz0;
    
    
%     % Time-stepping loop
%     for i = 1:length(t)-1
%         % Compute accelerations from state-space equations
%         
%         v_dot(i,1) = (g * p(i,3) / L^2) * p(i,1) - b*p_dot(i,1);
%         v_dot(i,2) = (g * p(i,3) / L^2) * p(i,2) - b*p_dot(i,2);
%         v_dot(i,3) = -g +(g * (p(i,3)*p(i,3)) / L^2) - b*p_dot(i,3);
%         
%         % Update velocities 
%         p_dot(i+1,:) = p_dot(i,:) + v_dot(i,:)*dt;
%         
%         % Update positions 
%         p(i+1,:) = p(i,:) +  p_dot(i+1,:)*dt; 
%         
%     end
 

    % Time-stepping loop
    for i = 1:length(t)-1
      
        Np = eye(3) - p(i,:)'*p(i,:)/(norm(p(i,:))^2);
        % Compute accelerations from state-space equations
        
        v_dot(i,:) = (Np*[0; 0; -g])' - (b/m)*p_dot(i,:);
        
        % Update velocities 
        
        p_dot(i+1,:) = p_dot(i,:) + v_dot(i,:)*dt;
        
        % Update positions 
        p(i+1,:) = p(i,:) + (Np*p_dot(i+1,:)')'*dt; 
        
    end

state = [p(:,1),p(:,2),p(:,3),p_dot(:,1),p_dot(:,2),p_dot(:,3)];
end

File: simulations/simulations_python/Pendulum.py
import numpy as np

g = 9.81  # gravitational acceleration (m/s^2)
L = 1    # length of the pendulum (m)
b = 0.04
m = 1
dt = 0.033



def Pendulum(x_pos0, y_pos0, z_pos0, vx0, vy0, vz0):
    # Time parameters
    t_end = 3      # End time for simulation
    t = np.arange(0, t_end, dt)  # Time vector

    # Preallocate arrays for storing results
    v_dot = np.zeros((len(t), 3))
    p_dot = np.zeros((len(t), 3))
    p = np.zeros((len(t), 3))

    # Set initial conditions
    p[0, 0] = x_pos0
    p[0, 1] = y_pos0
    p[0, 2] = z_pos0
    p_dot[0, 0] = vx0
    p_dot[0, 1] = vy0
    p_dot[0, 2] = vz0

    # Time-stepping loop
    for i in range(len(t) - 1):
        Np = np.eye(3) - np.outer(p[i, :], p[i, :]) / (np.linalg.norm(p[i, :]) ** 2)
        # Compute accelerations from state-space equations
        v_dot[i, :] = (Np @ np.array([0, 0, -g])) - (b / m) * p_dot[i, :]

        # Update velocities
        p_dot[i + 1, :] = p_dot[i, :] + v_dot[i, :] * dt

        # Update positions
        p[i + 1, :] = p[i, :] + (Np @ p_dot[i + 1, :]) * dt

    state = np.array(np.hstack((p, p_dot)))
    return state
